core/lib/dal/src/contract_verification_dal.rs

```diff
@@ -560,6 +587,285 @@ impl ContractVerificationDal<'_, '_> {
+        })
+        .instrument("get_contract_verification_info_v2#perfect_match")
+        .with_arg("address", &address)
+        .fetch_optional(self.storage)
+        .await?
+        .flatten()
+        {
+            return Ok(Some(info));
+        }
+
+        // Try partial match
+        let Some(deployed_contract) = self.get_contract_info_for_verification(address).await?
+        else {
+            return Ok(None);
+        };
+
+        let bytecode_marker = BytecodeMarker::new(deployed_contract.bytecode_hash)
+            .context("unknown bytecode kind")?;
+        let deployed_bytecode = match bytecode_marker {
+            BytecodeMarker::EraVm => deployed_contract.bytecode.as_slice(),
+            BytecodeMarker::Evm => trim_padded_evm_bytecode(
+                BytecodeHash::try_from(deployed_contract.bytecode_hash)
+                    .context("Invalid bytecode hash")?,
+                &deployed_contract.bytecode,
+            )
+            .context("invalid stored EVM bytecode")?,
+        };
+
+        let identifier = ContractIdentifier::from_bytecode(bytecode_marker, deployed_bytecode);
+        // `unwrap_or_default` is safe, since we're checking that `bytecode_without_metadata_keccak256` is not null.
+        let bytecode_without_metadata_keccak256 = identifier
+            .bytecode_without_metadata_keccak256
+            .as_ref()
+            .map(|h| h.hash())
+            .unwrap_or_default();
+
+        let Some((mut info, bytecode_keccak256, bytecode_without_metadata_keccak256)) =
+            sqlx::query!(
+                r#"
+                SELECT
+                    verification_info,
+                    bytecode_keccak256,
+                    bytecode_without_metadata_keccak256
+                FROM
+                    contract_verification_info_v2
+                WHERE
+                    bytecode_keccak256 = $1
+                    OR
+                    (
+                        bytecode_without_metadata_keccak256 IS NOT null
+                        AND bytecode_without_metadata_keccak256 = $2
+                    )
+                "#,
+                identifier.bytecode_keccak256.as_bytes(),
+                bytecode_without_metadata_keccak256.as_bytes()
+            )
+            .try_map(|row| {
+                let info = serde_json::from_value::<VerificationInfo>(row.verification_info)
+                    .decode_column("verification_info")?;
+                let bytecode_keccak256 = H256::from_slice(&row.bytecode_keccak256);
+                let bytecode_without_metadata_keccak256 = row
+                    .bytecode_without_metadata_keccak256
+                    .map(|h| H256::from_slice(&h));
+                Ok((
+                    info,
+                    bytecode_keccak256,
+                    bytecode_without_metadata_keccak256,
+                ))
+            })
+            .instrument("get_contract_verification_info_v2#perfect_match")
```

Comment:
Nit: previous query is instrumented with the same tag and I think it's misleading as it's not always "perfect match". Let's rename first one to "#direct" and this one to "#similar" or smth else.